Layers endpoint description (#141)
resource-watch/doc-api: source/includes/api/_layer.md

# Layer

## What is a layer?

By now, you are probably already familiar with [datasets](#dataset) and [querying](#query) them (if you are not, now is a good time to get up to speed on those). Many of the datasets you'll find on the RW API - and perhaps the datasets you are uploading too - contain georeferenced information. If that's the case, then you may want to render your data as a web map layer, and the RW API's *layer* endpoints can help you with that.

As we've seen in the [layer concept docs](#layer), a RW API layer may store data in different formats, depending on the needs of its author. This is done using the several open format fields a layer has. To keep this documentation easy to understand, we'll spit our approach to layers into two sections:

- We'll first discuss the details of the endpoints that allow you to [CRUD](https://en.wikipedia.org/wiki/Create,_read,_update_and_delete) layers, without focusing on the actual data contained in it.

> - In an upcoming section, we'll dive deeper into some of the most common structures used to store data in the layer's open format fields, and help you use that data to render an actual layer on a web application.

Reviewer: Assuming this refers to the tutorials that Logan is writing, I think this would fit best in a separate Tutorials section. I will make a task is Jira to remind us to fix the link.
Also, would change "upcoming section" to "future update" - "upcoming" makes it seem like I'll find it if I scroll down.

Author: :+1:

```shell
curl -X POST "https://api.resourcewatch.org/layer/find-by-ids" \
-H "Authorization: Bearer <your-token>" \
-H 'Content-Type: application/json' \
-d '{
	"ids": ["<dataset 1 id>", "<dataset 2 id>"],
    "app" "rw,prep"
}'
```

> Return all layers associated with two datasets, that are associated with both `rw` and `prep` applications simultaneously 

```shell
curl -X POST "https://api.resourcewatch.org/layer/find-by-ids" \
-H "Authorization: Bearer <your-token>" \
-H 'Content-Type: application/json' \
-d '{
	"ids": ["<dataset 1 id>", "<dataset 2 id>"],
    "app" "rw@prep"
}'
```

Besides the required `ids` array, your request body may optionally include a `app` string value if you'd like to filter the returned layers by their `application`:

- Use a single value, like `rw`, if you want to show only layers that have `rw` as one of their applications. 
- Use a comma separated list, like `rw,prep`, if you want to show only layers that have `rw` or `prep` as one of their applications.
- Use a @ separated list, like `rw@prep`, if you want to show only layers that have both `rw` and `prep` as their applications.

> - None of the filters require exact matching - a layer that simultaneously contain the applications `rw`, `prep` and `gfw` would match all 3 filters above.

Reviewer: This bullet doesn't make a ton of sense to me...

Author: what exactly is not clear? Should we work on the format, or on the message?

Reviewer: I just didn't understand what the last bullet actually means. Now that I've read it a few times I get it. I will suggest some wording that might clear it up.

Reviewer: ```suggestion
- Note that the the filters do not need to match on the full application list. For example, the filters `rw,prep` and `re@prep` will both match a layer with the application list `["rw", "prep", "gfw"]`.
```

> Sorting layers

```shell
curl -X GET "https://api.resourcewatch.org/v1/layer?sort=name"
```

Multiple sorting criteria can be used, separating them by commas.

> Sorting layers by multiple criteria

```shell
curl -X GET "https://api.resourcewatch.org/v1/layer?sort=name,slug"
```

> You can specify the sorting order by prepending the criteria with either `-` or `+`. By default, `asc` order is assumed.

Reviewer: Which one is which?

> Explicit order of sorting

```shell
curl -X GET "https://api.resourcewatch.org/v1/layer?sort=-name,+slug"
```

> Sorting layers by the role of the user who owns the layer

```shell
curl -X GET "https://api.resourcewatch.org/v1/layer?sort=user.role"
```

### Include related entities

> When fetching layers, you can request additional entities to be loaded. The following entities are available:

Reviewer: How consistent is the `includes` implementation? We might want to consider breaking that out into its own section at the top and then just referencing it for each resource that supports it.

Author: not a lot, so far widgets, layers and datasets support different combinations of includes, so having a central section would potentially make it confusing to the cover what is supported where.

               "account": "rw",
               "body": {
                   "maxzoom": 18,
                   "minzoom": 3,
                   "layers": [
                       {
                           "type": "mapnik",
                           "options": {
                               "sql": "SELECT * cait_2_0_country_ghg_emissions_filtered",
                               "cartocss": "",
                               "cartocss_version": "2.3.0"
                           }
                       }
                   ]
               }
           },
           "legendConfig": {
               "marks": {
                   "type": "rect",
                   "from": {
                       "data": "table"
                   }
               }
           },
           "applicationConfig": {},
           "staticImageConfig": {}
       }
   },
   "meta": {
       "status": "saved",
       "published": true,
       "updatedAt": "2017-01-23T16:51:42.571Z",
       "createdAt": "2017-01-23T16:51:42.571Z"
   }
}
```

If you know the id or the `slug` of a layer, then you can access it directly. Both id and `slug` are case-sensitive.

### Include related entities

> You can load related `user` and `vocabulary` data in the same request. See [this section](#include-related-entities135) for more details.

Reviewer: Looks like I jumped the gun on my previous comment, although this link doesn't work.

Author: If I understood correctly, that's a different thing:
- your previous comment was about having a common "includes" section that would be shared by dataset, widgets, layers, etc etc etc. While the idea is good, the fact that different resources support different includes makes me think it's implementation may be confusing to the end user, as I commented above
- This link (which is indeed broken, and is putting up a fight) points to the "includes" from the same layer resource - it basically means your get single layer request supports the same includes as you get multiple layers request.

I'll keep you up to date on the link bug

Author: update on link bug: this seems to be an issue with slate, the tool we use to build the html from the markdown files. The only solution I could find was to slightly change the titles of the sections, so that they don't repeat.

- Depending on your user account's role, you may have permission to create a layer but not delete it afterwards.
- The layers you create on the RW API will be publicly visible and available to other users.

Creating a layer is done using a POST request and passing the relevant data as body files. The supported body fields are as defined on the [layer reference](#layer-reference) section, but the minimum field list you must specify for all layers is:

- name
- description 
- application

There's also a dependency on a dataset id, as it is required to build the POST URL. As noted on the [layer concept](#layer) documentation, a layer is meant to hold the rendering details of a layer, but not the actual data - that should be part of the dataset. While this is not enforced - it's up to your rendering tool to load the data, and it can do it from a RW API dataset of from anywhere else - it's common and best practice to have the data for a layer be sourced from the dataset that's associated with it.

When a layer is created, a [vocabulary tag](#vocabulary-and-tags) for it is automatically created, associated with the dataset tag.

> The layer service was built to be very flexible, and not be restricted to specific layer rendering libraries or tools. This gives you the freedom to use virtually any rendering technology you want, but it also means you'll have to make additional decisions on how to structure your data into the different open format fields provided by a RW API layer. Further below we'll show you some examples of how existing applications use different rendering tools with the layer endpoints, to give you an idea on how you can structure your own data, and also as a way to help you get started creating your first layers for your own custom applications.

Reviewer: Again, I would change this to "in the future" and I'll note in Jira that we need to change it when the tutorials are in place.

Author: :+1:

provider                | String         | No                   |                            | Layer provider. It typically identifies the source service for the data displayed in the layer.
type                    | String         | No                   |                            | Layer type.
userId                  | String         | Yes (autopopulated)  |                            | Id of the user who created the layer. Set automatically on creation. Cannot be modified by users.
default                 | Boolean        | No                   | false                      | If the layer should be used as the dataset's default layer.
protected               | Boolean        | Yes                  | false                      | If the layer is protected. A protected layer cannot be deleted.               
published               | Boolean        | Yes                  | true                       | If the layer is published or not.                                               
env                     | String         | Yes                  | production                 | Environment to which the layer belongs.

> applicationConfig       | Object         | No                   |                            | Open schema object meant to host application-specific data or behavior configuration.

Reviewer: I would not recommend the phrase "Open schema" because it makes me think of Open Graph and schema.org, two competing web tagging standards. But I am open to the possibility that it's just me.

Author: I'm happy to change it to something else, but I can't put my finger on a better wording. Something like "open format" was the first thing that came to mind, but format, to me, sounds more like JSON vs XML/TXT/etc, which is not what we have here - the format is JSON, and it's not configurable. 
Any suggestions from your end?

Reviewer: Maybe "schema-less object"?

            "application": [
                "rw"
            ],
            "iso": [],
            "userId": "5dbadb06df2dc74d2ad054fb",
            "default": false,
            "protected": false,
            "published": true,
            "env": "production",
            "layerConfig": {},
            "legendConfig": {},
            "interactionConfig": {},
            "applicationConfig": {},
            "staticImageConfig": {},
            "createdAt": "2020-06-04T14:28:24.575Z",
            "updatedAt": "2020-06-04T14:28:24.575Z"
        }
    }
}
```

In this section we'll guide you through the process of creating a layer in the RW API. Layer creation is available to all registered API users, and will allow you to store your own layer visualisation settings on the API, for reusing and sharing.

Before creating a layer, there are a few things you must know and do:

- In order to be able to create a layer, you need to be [authenticated](#authentication).
- Depending on your user account's role, you may have permission to create a layer but not delete it afterwards.
- The layers you create on the RW API will be publicly visible and available to other users.

> Creating a layer is done using a POST request and passing the relevant data as body files. The supported body fields are as defined on the [layer reference](#layer-reference) section, but the minimum field list you must specify for all layers is:

Reviewer: ```suggestion
Creating a layer is done using a POST request and passing the relevant data as body fields. The supported body fields are as defined on the [layer reference](#layer-reference) section, but the minimum field list you must specify for all layers is:
```

- name
- description 
- application

> There's also a dependency on a dataset id, as it is required to build the POST URL. As noted on the [layer concept](#layer) documentation, a layer is meant to hold the rendering details of a layer, but not the actual data - that should be part of the dataset. While this is not enforced - it's up to your rendering tool to load the data, and it can do it from a RW API dataset of from anywhere else - it's common and best practice to have the data for a layer be sourced from the dataset that's associated with it.

Reviewer: ```suggestion
There's also a dependency on a dataset id, as it is required to build the POST URL. As noted on the [layer concept](#layer) documentation, a layer is meant to hold the rendering details of a layer, but not the actual data - that should be part of the dataset. While this is not enforced - it's up to your rendering tool to load the data, and it can do it from a RW API dataset or from anywhere else - it's common and best practice to have the data for a layer be sourced from the dataset that's associated with it.
```
